lib/src/globals.rs

```diff
@@ -0,0 +1,58 @@
+//! Global functions.
+
+use super::Result;
+use ostree::glib;
+use std::fs::File;
+use std::path::{Path, PathBuf};
+
+struct ConfigPaths {
+    persistent: PathBuf,
+    runtime: PathBuf,
+}
+
+/// Get the runtime and persistent config directories.  In the system (root) case, these
+/// system(root) case:  /run/ostree           /etc/ostree
+/// user(nonroot) case: /run/user/$uid/ostree ~/.config/ostree
+fn get_config_paths() -> ConfigPaths {
+    let mut r = if rustix::process::getuid() == rustix::process::Uid::ROOT {
+        ConfigPaths {
+            persistent: PathBuf::from("/etc"),
+            runtime: PathBuf::from("/run"),
+        }
+    } else {
+        ConfigPaths {
+            persistent: glib::user_config_dir(),
+            runtime: glib::user_runtime_dir(),
+        }
+    };
+    let path = "ostree";
+    r.persistent.push(path);
+    r.runtime.push(path);
+    r
+}
+
+impl ConfigPaths {
+    /// Return the path and an open fd for a config file, if it exists.
+    pub(crate) fn read_file(&self, p: impl AsRef<Path>) -> Result<Option<(PathBuf, File)>> {
+        let p = p.as_ref();
+        let mut runtime = self.runtime.clone();
+        runtime.push(p);
+        if let Some(f) = crate::container_utils::open_optional(&runtime)? {
+            return Ok(Some((runtime, f)));
+        }
+        let mut persistent = self.persistent.clone();
+        persistent.push(p);
+        if let Some(f) = crate::container_utils::open_optional(&persistent)? {
+            return Ok(Some((persistent, f)));
+        }
+        Ok(None)
+    }
+}
+
+/// Return the path to the global container authentication file, if it exists.
+pub(crate) fn get_global_authfile_path() -> Result<Option<PathBuf>> {
+    let paths = get_config_paths();
```

Comment:
Not sure how you'd do it in Rust but `paths` feels like it should be C++ `static` since `glib::user_config_dir()` can't change at runtime. And maybe the same thing for `auth.json`, but I'm not sure if you want to support updating that at runtime? If not I was wondering if it could be part of a `ConfigPaths` singleton or something, but I haven't thought that through.

Reply:
Yep, makes sense, I switched to a [OnceCell](https://docs.rs/once_cell/1.9.0/once_cell/#lazy-initialized-global-data).

Reply:
*reads docs* ... :exploding_head: